Acquire GIL when constructing/destructing ConcretePyObjectHolder

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37311 Add then(callback)->Future API to utils/Future.h
* **#37870 Acquire GIL when constructing/destructing ConcretePyObjectHolder**



Differential Revision: [D21410785](https://our.internmc.facebook.com/intern/diff/D21410785)

Commit:
Acquire GIL when destructing ConcretePyObjectHolder

[ghstack-poisoned]

Commit:
Update on "Acquire GIL when destructing ConcretePyObjectHolder"


Differential Revision: [D21410785](https://our.internmc.facebook.com/intern/diff/D21410785)

[ghstack-poisoned]

mrshenli commented:
Hey @wanchaol, does the change look reasonable to you?

xush6528 commented on torch/csrc/jit/python/python_ivalue.h:
```diff
@@ -11,17 +11,21 @@ namespace ivalue {
+  ~ConcretePyObjectHolder() {
+    pybind11::gil_scoped_acquire ag;
+    py_obj_ = py::none();
+  }
```
I guess with this, we don't need the Future Deleter helper?

mrshenli replied:
Ah, right, we should be able to drop that lock.

xush6528 commented on torch/csrc/jit/python/python_ivalue.h:
```diff
@@ -11,17 +11,21 @@ namespace ivalue {
   static c10::intrusive_ptr<PyObjectHolder> create(py::object py_obj) {
-    return c10::make_intrusive<ConcretePyObjectHolder>(py_obj);
+    return c10::make_intrusive<ConcretePyObjectHolder>(std::move(py_obj));
   }
```
Before this change, I think when creating a PyObjectHolder, the passed py obj ref count +2 and then -1. this makes sure that it only +1 one time.

xush6528 commented on torch/csrc/jit/python/python_ivalue.h:
```diff
@@ -11,17 +11,21 @@ namespace ivalue {
 struct C10_EXPORT ConcretePyObjectHolder final : PyObjectHolder {
  public:
   static c10::intrusive_ptr<PyObjectHolder> create(py::object py_obj) {
```
Since this data wrapper takes care of acquiring GIL on destruction.

I wonder can it also acquiring GIL on creation?

```cpp
static c10::intrusive_ptr<PyObjectHolder> create(const py::object& py_obj) {
py::gil_scoped_acquire ag;
return c10::make_intrusive<ConcretePyObjectHolder>(py_obj);
}
```

mrshenli replied:
good point! I will make the change if this looks good to @wanchaol too.

Commit:
Update on "Acquire GIL when destructing ConcretePyObjectHolder"


Differential Revision: [D21410785](https://our.internmc.facebook.com/intern/diff/D21410785)

[ghstack-poisoned]

Commit:
Update on "Acquire GIL when constructing/destructing ConcretePyObjectHolder"




Differential Revision: [D21410785](https://our.internmc.facebook.com/intern/diff/D21410785)

[ghstack-poisoned]

Commit:
Update on "Acquire GIL when constructing/destructing ConcretePyObjectHolder"




Differential Revision: [D21410785](https://our.internmc.facebook.com/intern/diff/D21410785)

[ghstack-poisoned]

wanchaol commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -604,11 +604,15 @@ inline IValue toIValue(
-          obj.cast<py::object>());
-
+      py::object py_obj;
+      {
```
nit: the new scope seems not needed here.

wanchaol commented on torch/csrc/jit/python/python_ivalue.h:
```diff
@@ -11,17 +11,21 @@ namespace ivalue {
  public:
   static c10::intrusive_ptr<PyObjectHolder> create(py::object py_obj) {
-    return c10::make_intrusive<ConcretePyObjectHolder>(py_obj);
+    return c10::make_intrusive<ConcretePyObjectHolder>(std::move(py_obj));
```
can we move the gil acquisition to the constructor as well (i.e. move the logic from toIValue to here)?

xush6528 replied:
I guess it's to change the constructor takes py::handle, and cast it to py::object with GIL

wanchaol replied:
Yes :)

wanchaol replied:
Actually, I would prefer adding a new constructor that takes a `py::handle`,  and move the logic to there. We should still keep the constructor that takes `py::object` without acquire gil for C++ usage.

wanchaol reviewed (approved):
Some nits, generally looks good to me, thanks!

Commit:
Update on "Acquire GIL when constructing/destructing ConcretePyObjectHolder"




Differential Revision: [D21410785](https://our.internmc.facebook.com/intern/diff/D21410785)

[ghstack-poisoned]

Commit:
Update on "Acquire GIL when constructing/destructing ConcretePyObjectHolder"




Differential Revision: [D21410785](https://our.internmc.facebook.com/intern/diff/D21410785)

[ghstack-poisoned]
